lib/utils.ts

```diff
@@ -1,10 +1,228 @@
 import crypto from "crypto";
 import { LogLine } from "../types/log";
+import { TextAnnotation } from "../types/textannotation";
 
 export function generateId(operation: string) {
   return crypto.createHash("sha256").update(operation).digest("hex");
 }
 
+export function formatText(textAnnotations: TextAnnotation[]): string {
+  // cluster tokens by line
+  const lineCluster: Map<number, TextAnnotation[]> = new Map();
+  const lineThreshold = 10;
+
+  // sort annotations by their Y-coordinate
+  const sortedAnnotations = textAnnotations.sort((a, b) => a.midpoint.y - b.midpoint.y);
+
+  sortedAnnotations.forEach((annotation) => {
+    const lineY = annotation.midpoint.y;
+    const keys = Array.from(lineCluster.keys());
+    const lastY = keys[keys.length - 1];
+
+    if (keys.length > 0 && Math.abs(lineY - lastY) < lineThreshold) {
+      lineCluster.get(lastY)!.push(annotation);
+    } else {
+      lineCluster.set(lineY, [annotation]);
+    }
+  });
+
+  let canvasWidth = 80;
+
+  // calculate canvas width based on the longest line
```

Comment:
what longest line? again, some more general comments here would be super helpful to understand what's going on